File: v1/src/main/java/com/api/v1/services/AllRentsRetrievalServiceImpl.java
package com.api.v1.services;

import com.api.v1.domain.RentRepository;
import com.api.v1.dtos.RentResponseDto;
import com.api.v1.mappers.RentResponseMapper;
import com.api.v1.utils.RentFinderUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
class AllRentsRetrievalServiceImpl implements AllRentsRetrievalService {

    @Autowired
    private RentRepository repository;

    @Autowired
    private RentFinderUtil rentFinderUtil;

    @Override
    public Mono<RentResponseDto> findByOrderNumber(String orderNumber) {
        return rentFinderUtil
                .find(orderNumber)
                .flatMap(rent -> Mono.just(new RentResponseMapper(
                        rent,
                        rent.getCustomer(),
                        rent.getCar(),
                        rent.getPayment()
                ).map()));
    }

    @Override
    public Flux<RentResponseDto> findAll() {
        return repository
                .findAll()
                .flatMap(rent -> Mono.just(new RentResponseMapper(
                        rent,
                        rent.getCustomer(),
                        rent.getCar(),
                        rent.getPayment()
                ).map()));
    }

}

File: v2/src/main/kotlin/com/api/v2/rent/service/AllRentsRetrievalServiceImpl.kt
package com.api.v2.rent.service

import com.api.v2.rent.dtos.RentResponseDto
import com.api.v2.rent.mappers.RentResponseMapper
import com.api.v2.rent.domain.RentRepository
import com.api.v2.rent.utils.RentFinderUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class AllRentsRetrievalServiceImpl: AllRentsRetrievalService {

    @Autowired
    lateinit var rentFinderUtil: RentFinderUtil

    @Autowired
    lateinit var rentRepository: RentRepository

    override suspend fun findByOrderNumber(orderNumber: String): RentResponseDto {
        return withContext(Dispatchers.IO) {
            val existingRent = rentFinderUtil.find(orderNumber)
            RentResponseMapper.map(
                existingRent,
                existingRent.customer,
                existingRent.car,
                existingRent.payment
            )
        }
    }

    override suspend fun findAll(): Flow<RentResponseDto> {
        return withContext(Dispatchers.IO) {
            rentRepository
                .findAll()
                .map{ existingRent ->
                    RentResponseMapper.map(
                        existingRent,
                        existingRent.customer,
                        existingRent.car,
                        existingRent.payment
                    )
                }
        }
    }

}